working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+not among the types that are implemented by the on-type of the extension.
+
+A compile-time error occurs if a member included by the show/hide part has
+a basename which is also the basename of a member declaration in the
+extension type.
+
+*For instance, if an extension `E` contains a declaration of a method named
+`toString`, the hide clause must include `toString` (or a class type,
+because they all include `toString`). Otherwise, the member declaration
+named `toString` would be an error.*
+
+Let `E` be an extension type with a show/hide part such that a member `m`
+is included in the interface of `E`. The member signature of `m` is the
+member signature of `m` in the on-type of `E`.
+
+A type in a hide or show clause may be raw (*that is, an identifier or
+qualified identifier denoting a generic type, but no actual type
+arguments*). In this case the omitted type arguments are determined by the
+corresponding superinterface of the on-type.
+
+*For example:*
+
+```dart
+extension type MyInt on int show num, isEven hide floor {
+  int get twice => 2 * this;
+}
+
+void main() {
+  MyInt m = 42;
+  m.twice; // OK, in the extension type.
+  m.isEven; // OK, a shown instance member.
+  m.ceil(); // OK, a shown instance member.
+  m.toString(); // OK, an `Object?` member.
+  m.floor(); // Error, hidden.
+}
+```
+
+
+### Implementing superinterfaces
+
+This section specifies the effect of having an `<interfaces>` part in an
+extension declaration.
+
+Let `E` be an extension declaration where `<interfaces>?` is of the form
+`implements T1, .. Tm`. We say that `E` has `T1` .. `Tm` as its direct
+superinterfaces.
+
+A compile-time error occurs if a direct superinterface does not denote a
+class, or if it denotes a class which cannot be a superinterface of a
+class.
+
+*For instance, `implements int` is an error.*
+
+If the `<interfaces>?` part of `E` is empty, it is treated as
+`implements Object`.
+
+For each member `m` named `n` in each direct superinterface of `E`, an
+error occurs unless `E` declares a member `m1` named `n` which is a correct
+override of `m`, or the show/hide part of `E` enables an instance member of
+the on-type which is a correct override of `m`.
+
+No subtype relationship exists between `E` and `T1, .. Tm`.
+
+*This means that when an extension type implements a set of interfaces, it
+is enforced that all the specified members are available, and that they
+have a signature which is compatible with the ones in `T1, .. Tm`, but
+there is no assignability from an expression of type `E` to a variable
+whose declared type is `Tj` for some `j` in 1..m. For that, it is necessary
+to use `box`, as described below.*
+
+
+### Gaining control over the instances: Protected extension types
+
+This section specifies the effect of including the keyword `protected` as
+the first token in an extension declaration.
+
+*The core idea is that no object can have a protected extension type as its
```

Comment:
This isn't true, as best I can tell.  Nothing stops me from writing this code, as far as I can see:

```dart
protected extension type Even on int {
  Even(int i) {
     if (i.isEven) return i;
     throw "Bad int, not even";
  }
}

void main() {
  List<Even> evens = [];
  List<Object> objects = evens;
  objects.add(3);
   print(evens[0]); // Prints 3
}
```

Nothing stops this code from running right?

Reply:
It wouldn't reach the print statement, because there would be a dynamic error at `objects.add(3)`: The type argument of `objects` is `Even` (which is reified as a type different from `int`), and the type check related to the parameter of `add` being covariant-by-class would amount to

```dart
  void add(E value) {
    value as E; // Implicitly induced type check.
  }
```

and that type cast would throw in the case where `E` is bound to `Even`:

If we include support for a `bool get verifyThis` then we would invoke `verifyThis` with `this` bound to the given `int`, and it would return false, so the cast would throw. If we don't include support for `verifyThis` then any cast to `E` would have to fail, unconditionally.

But if we have use-site invariance then we can do several things:

```dart
void main() {
  List<exactly Even> evens = [];
  // List<Object> objects = evens; // Compile-time error, so we don't do that.
  // evens.add(3); // Compile-time error: `3` does not have type `Even`.
  try { evens.add(Even(3)); } catch (_) {} // Does throw.
  evens.add(Even(4)); // OK.
}
```

Reply:
As discussed elsewhere, I think this is the one piece of this proposal that doesn't hang together yet, and I'm hesitant to land this without having something worked out, or at least acknowledging in the text that this doesn't hang together yet.